pkg/sfu/audiolevel.go

```diff
@@ -0,0 +1,91 @@
+package sfu
+
+import (
+	"sort"
+	"sync"
+)
+
+type audioStream struct {
+	id    string
+	sum   int
+	total int
+}
+
+type audioLevel struct {
+	sync.RWMutex
+	threshold uint8
+	streams   []*audioStream
+}
+
+func newAudioLevel(threshold uint8) *audioLevel {
+	if threshold > 127 {
+		threshold = 127
+	}
+	return &audioLevel{
+		threshold: threshold,
+	}
+}
+
+func (a *audioLevel) addStream(streamID string) {
+	a.Lock()
+	a.streams = append(a.streams, &audioStream{id: streamID})
+	a.Unlock()
+}
+
+func (a *audioLevel) removeStream(streamID string) {
+	a.Lock()
+	defer a.Unlock()
+	idx := -1
+	for i, s := range a.streams {
+		if s.id == streamID {
+			idx = i
+			break
+		}
+	}
+	if idx == -1 {
+		return
+	}
+	a.streams[idx] = a.streams[len(a.streams)-1]
+	a.streams[len(a.streams)-1] = nil
+	a.streams = a.streams[:len(a.streams)-1]
+}
+
+func (a *audioLevel) observe(streamID string, dBov uint8) {
+	a.RLock()
+	defer a.RUnlock()
+	for _, as := range a.streams {
+		if as.id == streamID {
+			if dBov <= a.threshold {
+				as.sum += int(dBov)
+				as.total += 1
+			}
+			return
+		}
+	}
+}
+
+func (a *audioLevel) calc() []string {
+	a.Lock()
+	defer a.Unlock()
+
+	sort.Slice(a.streams, func(i, j int) bool {
+		si, sj := a.streams[i], a.streams[j]
+		switch {
+		case si.total != sj.total:
+			return si.total > sj.total
+		default:
+			return si.sum < sj.sum
+		}
+	})
+
+	streamIDs := make([]string, len(a.streams))
+	for idx, s := range a.streams {
+		if s.total > 0 {
+			streamIDs[idx] = s.id
+		}
+		s.total = 0
+		s.sum = 0
+	}
+
+	return streamIDs
+}
```

Comment:
This will return all the streams for which we got a packet of noise since the last call, ordered by number of samples then by loudness. I don't think that will give session.go enough information to calculate if someone is speaking or not.

Reply:
I pulled out the audio parts of our code to a gist: https://gist.github.com/grahamking/4cd83cb22186d8a7e700918520dcd6be

Reply:
The number of samples which are over the threshold, so this will return all stream which get over the threshold where the first is the one with more packets over the threshold, if there are many with the same amount of packets over the threshold it will be ordered aslo by the loudest